src/System.Management.Automation/engine/hostifaces/History.cs

```diff
@@ -1017,137 +995,156 @@ public string Id
-                    // commands to always run in FullLanguage mode unless in a nested prompt.
-                    if (localRunspace != null)
-                    {
-                        localRunspace.InInternalNestedPrompt = ps.IsNested;
-                    }
-
-                    Collection<PSObject> results = ps.Invoke();
-                    if (results.Count > 0)
-                    {
-                        WriteObject(results, true);
-                    }
-
-                    pipeline.RemoveFromInvokeHistoryEntryList(entry);
+                    // Echo command
+                    Host.UI.WriteLine(commandToInvoke);
                 }
-                finally
+                catch (HostException)
                 {
-                    if (localRunspace != null)
+                    // when the host is not interactive, HostException is thrown
+                    // do nothing
+                }
+
+                // Items invoked as History should act as though they were submitted by the user - so should still come from
+                // the runspace itself. For this reason, it is insufficient to just use the InvokeScript method on the Cmdlet class.
+                using (System.Management.Automation.PowerShell ps = System.Management.Automation.PowerShell.Create(RunspaceMode.CurrentRunspace))
+                {
+                    ps.AddScript(commandToInvoke);
+
+                    EventHandler<DataAddedEventArgs> debugAdded = delegate (object sender, DataAddedEventArgs e)
+                    { DebugRecord record = (DebugRecord)((PSDataCollection<DebugRecord>)sender)[e.Index]; WriteDebug(record.Message); };
+                    EventHandler<DataAddedEventArgs> errorAdded = delegate (object sender, DataAddedEventArgs e)
+                    { ErrorRecord record = (ErrorRecord)((PSDataCollection<ErrorRecord>)sender)[e.Index]; WriteError(record); };
+                    EventHandler<DataAddedEventArgs> informationAdded = delegate (object sender, DataAddedEventArgs e)
+                    { InformationRecord record = (InformationRecord)((PSDataCollection<InformationRecord>)sender)[e.Index]; WriteInformation(record); };
+                    EventHandler<DataAddedEventArgs> progressAdded = delegate (object sender, DataAddedEventArgs e)
+                    { ProgressRecord record = (ProgressRecord)((PSDataCollection<ProgressRecord>)sender)[e.Index]; WriteProgress(record); };
+                    EventHandler<DataAddedEventArgs> verboseAdded = delegate (object sender, DataAddedEventArgs e)
+                    { VerboseRecord record = (VerboseRecord)((PSDataCollection<VerboseRecord>)sender)[e.Index]; WriteVerbose(record.Message); };
+                    EventHandler<DataAddedEventArgs> warningAdded = delegate (object sender, DataAddedEventArgs e)
+                    { WarningRecord record = (WarningRecord)((PSDataCollection<WarningRecord>)sender)[e.Index]; WriteWarning(record.Message); };
```

Comment:
```suggestion
                    EventHandler<DataAddedEventArgs> debugAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        DebugRecord record = (DebugRecord)((PSDataCollection<DebugRecord>)sender)[e.Index]; WriteDebug(record.Message);
                    };
                    EventHandler<DataAddedEventArgs> errorAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        ErrorRecord record = (ErrorRecord)((PSDataCollection<ErrorRecord>)sender)[e.Index]; WriteError(record);
                    };
                    EventHandler<DataAddedEventArgs> informationAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        InformationRecord record = (InformationRecord)((PSDataCollection<InformationRecord>)sender)[e.Index]; WriteInformation(record);
                    };
                    EventHandler<DataAddedEventArgs> progressAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        ProgressRecord record = (ProgressRecord)((PSDataCollection<ProgressRecord>)sender)[e.Index]; WriteProgress(record);
                    };
                    EventHandler<DataAddedEventArgs> verboseAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        VerboseRecord record = (VerboseRecord)((PSDataCollection<VerboseRecord>)sender)[e.Index]; WriteVerbose(record.Message);
                    };
                    EventHandler<DataAddedEventArgs> warningAdded = delegate (object sender, DataAddedEventArgs e)
                    {
                        WarningRecord record = (WarningRecord)((PSDataCollection<WarningRecord>)sender)[e.Index]; WriteWarning(record.Message);
                    };
```

@thomasrayner  I _think_ this is what it's looking for?

Reply:
@vexx32 looks like what I had in this commit which didn't build: https://github.com/PowerShell/PowerShell/commit/711a90e2ce90fc5caefa5c48856f3c640a038129

Reply:
@thomasrayner looks like you accidentally cut out one of the `delegate` lines in your version, which is why it didn't build. I've noticed mine could be a little better, I'll edit the suggestion.

Reply:
@vexx32 WELL THAT WOULD DO IT :joy: